### DISCUSSION TITLE
Bool ? Function()

### DISCUSSION THREAD

**Discussion by Monsoonexe** [dotnet/csharplang#5950] [Language Ideas]
Currently, C# supports the ternary operator and the null-checking shorthand (a = bool ? B : C) and (class?.Method()).

I propose something of the form `boolExpression ? Method();` This is essentially shorthand for `if (boolExpression) Method();`

It is clear and easy to read, and can allow for more one-line functions. 

`void MyFunc(bool enabled) => enabled ? SomeOtherFunction(args);`

Since `class?.Method()` boils down to `if (class != null) class.Method()`, where `class != null` is a boolean expression, this would mean we could make any boolean expression as the antecedent for a conditional function call.

**Comment by spydacarnage** [comment]
How is that really any better (and thus worth the development cost) than:
```cs
void MyFunc(bool enabled) { if(enabled) SomeOtherFunction(args); }
```

**Reply by FaustVX** [reply]
Your example doesn't work, the `=>` works only for expression, you have to use the full body method syntax.
```csharp
void MyFunc(bool enabled)
{
  if(enabled)
    SomeOtherFunction(args);
}
```

**Reply by Monsoonexe** [reply]
Exactly

**Comment by FaustVX** [comment]
your suggestion is very limited, it only works inside `void` returning method.
What is the result if `enabled` is `false` ? It's not clear.
using `enabled ? class.Method() : throw new Exception()` or `enabled ? class.Method() : null` is clearer.

**Reply by zms9110750** [reply]
with reason.
If everything is written in this way, maybe we won't need if in the future:
```
bool ? {code} : bool? {code} : {code}
```

**Reply by Monsoonexe** [reply]
In this suggestion, there is no 'else' like the ternary operator. So `bool ? {code}  : {code};` is not valid. In the same vein that there is no else-clause to `class?.Method();`

**Reply by Monsoonexe** [reply]
An example for a function that would use this outside of a quick and dirty one-line function (such as the void-returning function in the original post).

```
object SomeFunction(object someArg, bool someBool)
{
    SomePrecondition(someArg) ? throw new Exception();

    !initialized ? Initialize();
    
    object returnValue = Process(); //something arbitrary to get an object reference
  
    someBool ? ExtraProcessing(returnValue );

   return returnValue ;    
}
```

This wouldn't replace the if-else construct. It would reduce some of the boiler-plate involved in 
``` 
if (boolExpression) 
     Method();
```

**Reply by FaustVX** [reply]
The `class?.Method()` has an implicit `else` clause, which is `null`.
It's equivalent to `class is null ? class.Method() : null;`

**Comment by jnm2** [comment]
This is very reminiscent of https://github.com/dotnet/csharplang/discussions/5942 which has some issues, as discussed there.

**Reply by Monsoonexe** [reply]
The issues described in #5942 have to do with ambiguity with the else-clause. In this proposal, there is no else-clause (much in the same vein that there is no else-clause in `class?.Method();` I guess the implied else-clause is "don't call this function".

**Reply by jnm2** [reply]
`?.` isn't a comparable example in terms of ambiguity because there is a `.`. Whereas with `?` alone, it's not clear whether you're in a void context or not. The grammar itself can't fluctuate based on binding the symbols and finding the method and identifying it as not having a return value.
